PayloadPkg/FirmwareUpdate/FirmwareUpdate.c

```diff
@@ -1163,6 +1165,73 @@ InitFirmwareUpdate (
   return EFI_SUCCESS;
 }
 
+
+/**
+  Try to recover a working boot partition.
+
+  @retval  EFI_SUCCESS          The operation completed successfully.
+  @retval  others               There is error happening.
+**/
+EFI_STATUS
+InitFirmwareRecovery (
+  VOID
+)
+{
+  FLASH_MAP                   *FlashMap;
+  EFI_STATUS                  Status;
+  UINT32                      TopSwapRegionSize;
+  UINT32                      RedundantRegionSize;
+  UINT32                      TopSwapPrimaryAddress;
+  UINT32                      TopSwapBackupAddress;
+  UINT32                      RedundantPrimaryAddress;
+  UINT32                      RedundantBackupAddress;
+  BOOLEAN                     IsBackupUpdate;
+
+  FlashMap = GetFlashMapPtr ();
+  if (FlashMap == NULL) {
+    DEBUG((DEBUG_INFO, "Flash map not found!\n"));
+    return EFI_NOT_FOUND;
+  }
+
+  Status = GetRegionInfo (&TopSwapRegionSize, &RedundantRegionSize, NULL);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "GetRegionInfo, Status = 0x%x\n", Status));
+    return Status;
+  }
+
+  TopSwapPrimaryAddress = FlashMap->RomSize - TopSwapRegionSize;
+  TopSwapBackupAddress = TopSwapPrimaryAddress - TopSwapRegionSize;
+  RedundantPrimaryAddress = TopSwapBackupAddress - RedundantRegionSize;
+  RedundantBackupAddress = RedundantPrimaryAddress - RedundantRegionSize;
+
+  IsBackupUpdate = GetCurrentBootPartition() == 0;
+
+  DEBUG ((DEBUG_ERROR, "Copying top swap....\n"));
+  Status = CopyRegionBlock (TopSwapPrimaryAddress, TopSwapBackupAddress, TopSwapRegionSize, IsBackupUpdate);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "CopyRegionBlock, Status = 0x%x\n", Status));
+    return Status;
+  }
```

Comment:
Test with reset here to mock a power failure.

Reply:
Done. Current implementation will re-do the update on the next boot if a power failure occurs here.